docs/shim-rune.md: add design doc for shim-rune (#12)
confidential-containers/enclave-cc: docs/shim-rune.md

# Workflow

## 1. RunPodSandbox

Compared with `containerd-shim-runc-v2`, `shim-rune` should add the following functions when
launching the pod sandbox.

- In the `Create` API of shim-rune, shim-rune will create an agent enclave container with the
pre-created oci bundle after the pause container is created successfully.
- In the `Start` API of `shim-rune, shim-rune will start the agent enclave container after the pause
container is started successfully.

A pre-created oci bundle for the agent enclave container is needed instead of the form of container
image because the agent enclave container is the first component along with pod creation and there
is no direct support for it to pull and unpack it in a TEE. The `agent_container_instance` field in
the configuration file of shim-rune(In the Appendix section) shows the host path of oci bundle for
the agent enclave container. The contents of the bundle are as follows:

```sh
tree /opt/enclave-cc/agent-instance/ -L 1
/opt/enclave-cc/agent-instance/
├── config.json
└── rootfs
```

Note that oci bundle of the agent enclave container (`agent_container_instance` field in the

> shim-rune configuration file) must be **read-only** because the agent enclave containers of all

Author: Thank you for your suggestion. The config.json and rootfs in the bundle are read-only in our current design.

The pr-installed OCI bundle of agent enclave container will be shared on all pods in the host. Suppose the OCI bundle is in `/opt/enclave-cc/agent-instance/`. Using the same OCI bundle to run two agent enclave containers is impossible in Occlum Libos, because each occlum run will initialize the rootfs in the OCI bundle, and the second reinitialization will panic. 

After discussion, we provide the following solution:

Every time the agent container is created and started, `/opt/enclave-cc/agent-instance/` is used as the lower read-only layer, and then a read-write layer (upperdir) is superimposed through overlayfs, then the final OCI bundle for the agent container is aggregated.

As shown below, suppose the agent container  id is "a3e3c2"

```
mount -t overlay overlay -o
  lowerdir= /opt/enclave-cc/agent-instance/ ,
  upperdir=/run/containerd/agentenclave/a3e3c2/upper,
  workdir=/run/containerd/agentenclave/a3e3c2/work, 
  /run/containerd/agentenclave/a3e3c2/merged
```
Note that: `upperdir`, `mergeddir`, `workdir` are all created and managed by shim-rune.

> shim-rune configuration file) must be **read-only** because the agent enclave containers of all

Reviewer: makes sense

Pods on this node share this bundle. Once modified, it will affect the behavior of existing or newly
created agent enclave containers. Therefore, every time the agent enclave container is created,
`agent_container_instance` is used as the read-only layer and then overlays read and write layers

> to aggregate the final oci bundle for the agent enclave container.

Author: Sorry for the ambiguity,  `overlay read and write layers` is referred to `upperdir` in OverlayFS, which is read-write.

In addition, the oci bundle of agent container may have multiple different versions, they contain
different versions of untrusted PAL (even support different LibOS) and its dependencies, such as
configuration files. shim-rune just needs to call `rune` and pass the location of its oci bundle
to rune.

`rune` receives the path of the oci bundle passed by shim-rune, parses the oci container
configuration file, creates and enters a new mount namespace, and uses bind mount to mount the
rootfs directory tree in the oci bundle as the container rootfs. Finally, LibOS is loaded and

> initialized through the PAL API `pal_init()` as the app container's No. 1 process runelet. Please

Author: Thank you for your suggestions, I have modified them according to your suggestions and closed the conversion.

lower layers.
- In the short to medium term, the enclave-agent program cannot rely on the rootfs aggregation
capability of a specific LibOS to construct the oci bundle of the app container. shim-rune is
responsible for generating the bundle of the app container.

Considering these limitations and after discussing with Libos team, the first stage enclave-cc
uses the following methods to run the app enclave container.

### Occlum

Refer to the occlum guide of
[Runtime boot pre-generated UnionFS image](https://github.com/qzheng527/occlum/blob/combine_jsons/demos/runtime_boot).
A pre-created oci bundle for the boot instance is needed in the host. The boot instance is
responsible for using the customized `init`, mount and boot a pre-generated UnionFS image.

Considering the implementation of occlum dynamic mount image, shim-rune should overlay the
`boot instance bundle` and `encrypted unionfs image` to generate the final bundle of the app
container. For example:

```sh
mount -t overlay overlay -o
  lowerdir= <path of encrypted unionfs image>:<path of boot instance bundle>,
  upperdir=upper, workdir=work merged
```

> Then the question is where is the location of `boot instance bundle` and `encrypted unionfs image`?

Author: Hi, please let me explain how the agent enclave works first so that we can reach a consensus.
Due to the limitation of Occlum Libos, the implementation of agent-enclave is as follows:

1. When the shim-rune sends the `image_id` to the agent enclave container, the agent enclave container uses [image-rs](https://github.com/confidential-containers/image-rs) to pull the app image, and `image-rs` will unpack the N compressed layers into N directories under the `CC_IMAGE_WORK_DIR` directory.

2. The agent enclave program will first mount the fixed host directory into the enclave through the Occlum `mount` function. Please refer to [link](https://github.com/zhiwei-intel-h/image-rs/blob/dev-occlum-adaption/src/snapshots/unionfs.rs#L107).

3. The agent enclave program copies the contents of the N layers of the image to the fixed host directory. The content combination of N layers finally generates an encrypted unionfs (based on sefs) image on the host.

Such an implementation has a big performance problem. 

A possible solution is: 

- Libos (Occlum /Gramine) supports functions similar to overlay, avoids copying multiple layers, and supports building encrypted unionfs layer(sefs based) for each layer, so as to realize layer sharing between containers. In fact, the Occlum team is working on it.

> Then the question is where is the location of `boot instance bundle` and `encrypted unionfs image`?

Reviewer: > How the design described in this doc works when there are non-encrypted layers in the image?

I don't see this part answered. If we have a big performance problem with:

"The agent enclave program copies the contents of the N layers of the image to the fixed host directory"

why not only copy those layers that are encrypted (although it might be this does not help much if it's more common that the images have all the layers encrypted vs just a subset of them)

- Location of `boot instance bundle`

One host only needs a pre-created occlum oci bundle for the boot instance. The

> `boot_container_instance` field in shim configure file can show the location.

Author: Thank you, sounds great, we don't want to implement some Libos-specific functionality in shim-rune.

The pre-generated boot template instance looks like an oci bundle rootfs on the host side.
It contains the
[occlum boot template instance](https://github.com/qzheng527/occlum/blob/combine_jsons/demos/runtime_boot/README.md#build-a-boot-template-occlum-instance).

```sh

> tree /opt/enclave-cc/boot-instance -L 1

Reviewer 2: ```suggestion
$ tree /opt/enclave-cc/boot-instance -L 1
```

/opt/enclave-cc/boot-instance
└── rootfs

1 directory, 1 file
```

- Location of `encrypted unionfs image`

One pod only has one agent enclave container. The agent container needs to manage multiple images.
After the agent container pulls an image, it will convert all layers about this image into an
encrypted unionfs (based sefs) image in the host.

Since one image corresponds to a encrypted unionfs image, shim-rune can generate `container_id`
based on image name and pass `container_id` to agent enclave contianer with `PullImageRequest`.

> The agent enclave contianer can store the encrypted unionfs image to directory distinguished by

Reviewer 2: ```suggestion
The agent enclave container can store the encrypted unionfs image to directory distinguished by
```

`container_id` in host, such as `<agent_container_bundle_path>/rootfs/images/<container_id>/sefs`.

When shim-rune launch the app enclave container, shim-rune will find the corresponding encrypted
unionfs image through `container_id`. Then shim-rune overlay the `encrypted unionfs image` and
`boot instance bundle` to generate the final app enclave container bundle.

### Gramine

TODO

At last, shim-rune call `rune` to launch the app enclave container. The process is similar to
launch the agent enclave container with `rune`.

## 4. Stop sandbox

Compared with `containerd-shim-runc-v2`, shim-rune adds the work of kill/cleanup for the agent
enclave container.

# How to integrate with containerd

Because shim-rune supports containerd shim v2 API, you can add the associated configurations for
shim-rune in the containerd config file, e.g, `/etc/containerd/config.toml`, on your system.

```
        [plugins.cri.containerd]
          ...
          [plugins.cri.containerd.runtimes.rune]
            runtime_type = "io.containerd.rune.v2"
```
then restart containerd on your system.

# Appendix

## shim configuration sample

> ```

Reviewer 2: ```suggestion
```toml
```

log_level = "info" # "debug" "info" "warn" "error"

[containerd]
    socket = "/run/containerd/containerd.sock"
    agent_container_instance = "/opt/enclave-cc/agent-instance/"
    boot_container_instance = "/opt/enclave-cc/boot-instance/"
    agent_container_root_dir = "/run/containerd/agentenclave"
    agent_url = "tcp://0.0.0.0:7788"
```

where:
- @log_level: specify the log level for shim-rune.
- @socket: the containerd socket
- @agent_container_instance:  the host path of pre-created oci bundle for agent enclave container
- @boot_container_instance: the host path of pre-created oci bundle for boot instance
- @agent_container_root_dir: the root dir of agent enclave container running state
- @agent_url: the listening address of the agent container, through which the shim communicates
with the agent container to perform PullImage. (support unix and tcp socket communication)
  - Currently occlum 0.27 version does not support cross-world untrusted Unix domain socket.

>   - After occlum releases [NGO](https://github.com/occlum/ngo) at the end of June, the communication

Reviewer 2: Might be worth qualifying this as "June 2022"?

  based on tcp socket between shim-rune and agent container will be switched to ttrpc based on
  cross-world untrusted Unix domain socket.

# Reference

> - [kata container](https://github.com/confidential-containers/kata-containers-CCv0)

Reviewer 2: ```suggestion
- [Kata Containers](https://github.com/confidential-containers/kata-containers-CCv0)
```

> - [contained](https://github.com/confidential-containers/containerd)

Reviewer 2: ```suggestion
- [containerd](https://github.com/confidential-containers/containerd)
```

Author: Thank you very much for your suggestions, I have modified the pr as you suggested. Thank you very much!

3. The agent enclave store the encrypted unionfs image in host. We will discuss the
path of encrypted unionfs image in the next subsection(Create app container section).

### Communication protocol

Below is the communication protocol between shim-rune and agent enclave container, the
communication protocol is referred to
[kata containers](https://github.com/confidential-containers/kata-containers/blob/CCv0/src/libs/protocols/protos/image.proto).

```proto
syntax = "proto3";
...
// Image defines the public APIs for managing images.
service Image {
    // PullImage pulls an image with authentication config.
    rpc PullImage(PullImageRequest) returns (PullImageResponse) {}
}

message PullImageRequest {
    // Image name (e.g. docker.io/library/busybox:latest).
    string image = 1;
    // Unique image identifier, used to avoid duplication when unpacking the image layers.
    string container_id = 2;
    // Use USERNAME[:PASSWORD] for accessing the registry
    string source_creds = 3;
}

message PullImageResponse {
    // Reference to the image in use. For most runtimes, this should be an
    // image ID or digest.
    string image_ref = 1;
}
```

## 3. Create app container

> Ideally, just like `kata-agent`, the agent enclave container generates the bundle of the app
> container based on occlum or gramine. In this way, shim-rune can avoid generating the app container
> bundle which depends on specific Libos (occlum or gramine).

Reviewer 3: Wording fixup
```suggestion
Ideally, just like `kata-agent`, the agent enclave container generates the bundle of the app
container without requiring any confidential compute dependencies in the original container image. Unlike `kata-agent`, the agent enclave container will prepare the bundle in a way that a LibOS can load the application inside an enclave.
```

> But LibOS has limited capabilities such as:

Reviewer 3: Wording fixup
```suggestion
But each LibOS has limited capabilities such as:
```

responsible for using the customized `init`, mount and boot a pre-generated UnionFS image.

Considering the implementation of occlum dynamic mount image, shim-rune should overlay the
`boot instance bundle` and `encrypted unionfs image` to generate the final bundle of the app
container. For example:

```sh
mount -t overlay overlay -o
  lowerdir= <path of encrypted unionfs image>:<path of boot instance bundle>,
  upperdir=upper, workdir=work merged
```

Then the question is where is the location of `boot instance bundle` and `encrypted unionfs image`?

- Location of `boot instance bundle`

One host only needs a pre-installed occlum OCI bundle for the boot instance. The
`boot_container_instance` field in shim configure file can show the location.

The pre-generated boot template instance looks like an OCI bundle rootfs on the host side.
It contains the
[occlum boot template instance](https://github.com/qzheng527/occlum/blob/combine_jsons/demos/runtime_boot/README.md#build-a-boot-template-occlum-instance).

```sh
$ tree /opt/enclave-cc/boot-instance -L 1
/opt/enclave-cc/boot-instance
└── rootfs

1 directory, 1 file
```

- Location of `encrypted unionfs image`

> One pod only has one agent enclave container. The agent container needs to manage multiple images.

Reviewer 3: Wording fixup
```suggestion
Each pod has exactly one agent enclave container. The agent enclave container needs to manage multiple images.
```

> After the agent container pulls an image, it will convert all layers about this image into an

Reviewer 3: wording fixup
```suggestion
After the agent container pulls an image, it will convert all layers of this image into an
```

# Design

According to [Enclave Confidential Containers (Enclave-CC) Design and Architecture](./design.md),
shim-rune should launch the agent enclave container when pod creation. The agent enclave container
is deployed in the form of OCI bundle (pre-installed on each node) rather than a container image,
and thus `shim-rune` can call `rune` to start it. The agent enclave container has the same life
cycle as the pod. Its main job is to receive and then load the actual workload's container image
into an app enclave.

> ![image](./images/shim.png)

Reviewer 3: What do you want to communicate with the arrow from PAL to rune?
I think the other arrows loosely suggest call flow.

Author: Hi, thank you for your advice.
I have updated the [shim.png](https://github.com/haosanzi/enclave-cc/blob/main/docs/images/shim.png) to show the relationship with rune and pal/pal api.

PAL is the implementer of Enclave Runtime PAL API. [Enclave Runtime PAL API](https://github.com/inclavare-containers/inclavare-containers/blob/master/rune/libenclave/internal/runtime/pal/spec.md) defines a common interface to interact between rune and enclave runtime. `rune` manages the confidential container through the PAL API.

- [PAL API](https://github.com/inclavare-containers/inclavare-containers/blob/master/rune/libenclave/internal/runtime/pal/spec.md): This API defines the function calls between Enclave Runtime PAL and rune. 
- PAL: The implementer of Enclave Runtime PAL API, on behalf of Enclave Runtime. This [repo](https://github.com/occlum/occlum/tree/master/src/pal)  shows the implementation of occlum pal.
- Enclave Runtime: The implementer of enclave. Occlum, Graphene-SGX, and WAMR (WebAssembly Micro Runtime) are all the so-called Enclave Runtime. 

Please refer to [guide](https://github.com/inclavare-containers/inclavare-containers/blob/master/docs/design/terminology.md) for the terminology in inclavare-container.

As is shown in the picture, the binary name of `shim-rune` is `containerd-shim-rune-v2`.
`shim-rune` should implement the Containerd Runtime V2 (Shim API) for Enclave-CC. With `shim-rune`,
Kubernetes can launch Pod and OCI-compatible containers with one shim per Pod.

## Goals

- Implements [Containerd Runtime V2 (Shim API)](https://github.com/containerd/containerd/tree/main/runtime/v2)
- Use `rune` OCI runtime to manage container
- Launch agent enclave container during PodSandbox creation
- Complete PullImage with agent enclave container with ttrpc communication
- Launch application enclave container with FUSE encryption filesystem generated by agent enclave
container.
- Kill and cleanup agent enclave container resource when stopping PodSandbox

# Workflow

## 1. RunPodSandbox

Compared with `containerd-shim-runc-v2`, `shim-rune` should add the following functions when
launching the pod sandbox.

- In the `Create` API of shim-rune, shim-rune will create an agent enclave container with the
pre-installed OCI bundle after the pause container is created successfully.

> - In the `Start` API of `shim-rune, shim-rune will start the agent enclave container after the pause

Reviewer 3: missing back tick
```suggestion
- In the `Start` API of `shim-rune`, shim-rune will start the agent enclave container after the pause
```

`boot_container_instance` field in shim configure file can show the location.

The pre-generated boot template instance looks like an OCI bundle rootfs on the host side.
It contains the
[occlum boot template instance](https://github.com/qzheng527/occlum/blob/combine_jsons/demos/runtime_boot/README.md#build-a-boot-template-occlum-instance).

```sh
$ tree /opt/enclave-cc/boot-instance -L 1
/opt/enclave-cc/boot-instance
└── rootfs

1 directory, 1 file
```

- Location of `encrypted unionfs image`

Each pod has exactly one agent enclave container. The agent enclave container needs to manage
multiple images. After the agent container pulls an image, it will convert all layers of this
image into an encrypted unionfs (based sefs) image in the host.

Since one image corresponds to a encrypted unionfs image, shim-rune can generate `container_id`
based on image name and pass `container_id` to agent enclave container with `PullImageRequest`.
The agent enclave container can store the encrypted unionfs image to directory distinguished by
`container_id` in host, such as `<agent_container_bundle_path>/rootfs/images/<container_id>/sefs`.

> When shim-rune launch the app enclave container, shim-rune will find the corresponding encrypted

Reviewer 4: shim-rune --> rune launch the app enclave container
